File: Add Two Numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode ret = null;
        ListNode node = null;
        int carry = 0;
        while (l1 != null || l2 != null || carry > 0) {
            if (node == null) {
                ret = new ListNode(0);
                node = ret;
            }
            else {
                node.next = new ListNode(0);
                node = node.next;
            }
            int val1 = 0;
            if (l1 != null)
            {
                val1 = l1.val;
                l1 = l1.next;
            }
            int val2 = 0;
            if (l2 != null)
            {
                val2 = l2.val;
                l2 = l2.next;
            }
            int sum = val1 + val2 + carry;
            carry = sum / 10;
            int digit = sum % 10;
            node.val = digit;
        }
        return ret;
    }
}

File: Add Two Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        if (l1 == 0)
            return l2;
        ListNode* head = 0;
        ListNode* node = 0;
        bool carry = false;
        while (l1 || l2 || carry)
        {
            int digit = carry;
            if (l2 != 0)
            {
                digit += l2->val;
                l2 = l2->next;
            }
            if (l1 != 0)
            {
                digit += l1->val;
                l1 = l1->next;
            }
            carry = digit >= 10;
            digit %= 10;
            if (node)
            {
                node->next = new ListNode(digit);
                node = node->next;
            }
            else
            {
                node = new ListNode(digit);
                head = node;
            }
        }
        return head;
    }
};